### PR TITLE
bug(category): fix bug where multiple values of a particular filter were not being set properly

### PR DESCRIPTION
**Pull request by lderrickable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Filtering categories by two filters of the same kind (e.g. color: blue, red) doesn't work, because the filters are being built incorrectly.

Fixes: N/A

## What is the new behavior?
Filtering categories works properly.

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

## Other information
Additionally, I made daffodil filtering easier to use by removing the DaffCategoryFilterActionEnum and only requiring a filter type from the app dev. I also added some validation for the filters, and I added a model specifically for the DaffToggleCategoryFilter action, so there's no confusion for the end user needing to pass an array with only one value in it.

### CODE REVIEW

File: libs/category/src/models/requests/filter-action.ts [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
 
 /**
  * The separator between a range type filter's values
  * e.g. "price": "30-40"
  */
 export const DaffCategoryFromToFilterSeparator = '-';
 
-export interface DaffCategoryFilterAction {
-	action: DaffCategoryFilterActionEnum;
+export interface DaffToggledCategoryFilterAction {
```
Comment (damienwebdev) [inline libs/category/src/models/requests/filter-action.ts]: The trailing "Action" here is easily confused with a Redux action, let's suffix with `Request` instead.

File: libs/category/src/models/requests/filter-action.ts [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
+	type: DaffCategoryFilterType;
 	name: string;
 	value: string;
 	label?: string;
 }
+
+export interface DaffCategoryFilterAction {
```
Comment (damienwebdev) [inline libs/category/src/models/requests/filter-action.ts]: The trailing "Action" here is easily confused with a Redux action, let's suffix with `Request` instead.

Comment (damienwebdev) [inline libs/category/src/models/requests/filter-action.ts]: This should be a union type.

File: libs/category/src/effects/category.effects.ts [hunk 3]
```diff
@@ -143,18 +150,70 @@ export class DaffCategoryEffects {
...
+
+		return appliedFilters.concat([{
+			name: toggledFilter.name,
+			label: toggledFilter.label,
+			values: [toggledFilter.value],
+			type: toggledFilter.type
+		}]);
+	}
+	
+	validateFilters(filters: DaffCategoryFilterAction[]): void {
```
Comment (damienwebdev) [inline libs/category/src/effects/category.effects.ts]: This validation should be done as much as possible at compile time instead of runtime. I've left a comment on the model to attempt to help accomodate this.

File: libs/category/src/effects/category.effects.ts [hunk 4]
```diff
@@ -143,18 +150,70 @@ export class DaffCategoryEffects {
...
+		return appliedFilters.map(filter => {
+			if(filter.name === toggledFilter.name) {
+				filter.values = filter.values.filter(value => value !== toggledFilter.value);
+			}
+
+			return filter;
+		}).filter(filter => !!filter.values.length);
+	}
+
+	private addFilter(toggledFilter: DaffToggledCategoryFilterAction, appliedFilters: DaffCategoryFilterAction[])
```
Comment (damienwebdev) [inline libs/category/src/effects/category.effects.ts]: These look like reducer functions, they should be moved there.

File: libs/category/src/effects/category.effects.ts [hunk 5]
```diff
@@ -143,18 +150,70 @@ export class DaffCategoryEffects {
...
+
+	private isFilterApplied(toggledFilter: DaffToggledCategoryFilterAction, appliedFilters: DaffCategoryFilterAction[]): boolean {
+		return !!appliedFilters.find(filter => 
+			toggledFilter.name === filter.name && !!filter.values.filter(value => value === toggledFilter.value).length
+		);
 	}
 
-	private areFiltersEqual(filter1: DaffCategoryFilterAction, filter2: DaffCategoryFilterAction): boolean {
-		return filter1.name === filter2.name && filter1.value === filter2.value;
+	private removeFilter(toggledFilter: DaffToggledCategoryFilterAction, appliedFilters: DaffCategoryFilterAction[])
```
Comment (damienwebdev) [inline libs/category/src/effects/category.effects.ts]: These look like reducer functions, they should be moved there.

File: libs/category/src/effects/category.effects.ts [hunk 6]
```diff
@@ -143,18 +150,70 @@ export class DaffCategoryEffects {
 	)
 	
 	private toggleCategoryFilter(
```
Comment (damienwebdev) [inline libs/category/src/effects/category.effects.ts]: These look like reducer functions, they should be moved there.

**Comment by lderrickable** [comment]
@damienwebdev I didn't have time to finish the tests. Will do in a few hours. But everything should be addressed, and it works as expected.

File: libs/category/src/reducers/category/category.reducer.ts [hunk 7]
```diff
@@ -104,3 +116,96 @@ export function categoryReducer(state = initialState, action: DaffCategoryAction
...
+function addFilter(toggledFilter: DaffToggledCategoryFilterRequest, appliedFilters: DaffCategoryFilterRequest[])
+	: DaffCategoryFilterRequest[] {
+	return toggledFilter.type === DaffCategoryFilterType.Match ?
+		addMatchFilter(toggledFilter, appliedFilters) :
+		addEqualOrRangeFilter(toggledFilter, appliedFilters);
+}
+
+function addMatchFilter(toggledFilter: DaffToggledCategoryFilterRequest, appliedFilters: DaffCategoryFilterRequest[])
+: DaffCategoryFilterRequest[] {
+	return appliedFilters.concat([{
```
Comment (damienwebdev) [inline libs/category/src/reducers/category/category.reducer.ts]: This can be simplified to `appliesFilters.concat([toggledFilter])` I think.

File: libs/category/src/effects/category.effects.ts [hunk 8]
```diff
@@ -142,19 +162,21 @@ export class DaffCategoryEffects {
...
+			if(filter.type === DaffCategoryFilterType.Range &&
+				filter.value[0].split(DaffCategoryFromToFilterSeparator).length !== 2) {
+				throw new Error('Category range filter is in an invalid format. Should be **-**');
+			}
 
-	private areFiltersEqual(filter1: DaffCategoryFilterAction, filter2: DaffCategoryFilterAction): boolean {
-		return filter1.name === filter2.name && filter1.value === filter2.value;
+			for(let j=i+1; j < filters.length; j++) {
+				if(filters[i].name === filters[j].name) {
+					throw new Error('More than one category filter of the same name exists. These should' +
```
Comment (damienwebdev) [inline libs/category/src/effects/category.effects.ts]: This validation isn't necessary and work should be done in the reducer to accommodate this.

File: libs/cart/src/effects/cart.effects.ts [hunk 9]
```diff
@@ -55,12 +55,14 @@ export class DaffCartEffects<T extends DaffCart> {
   @Effect()
   get$ = this.actions$.pipe(
     ofType(DaffCartActionTypes.CartLoadAction),
```
Comment (damienwebdev) [inline libs/cart/src/effects/cart.effects.ts]: Why is this file changed?

File: libs/category/src/models/requests/filter-request.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import { DaffCategoryFilterType } from '../category-filter';
+
+/**
+ * The separator between a range type filter's values
+ * e.g. "price": "30-40"
+ */
+export const DaffCategoryFromToFilterSeparator = '-';
+
+export interface DaffToggledCategoryFilterRequest {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/category/src/models/requests/filter-request.ts]: What is the alternative?

File: libs/cart/src/effects/cart.effects.ts (hunk 9 above)
Comment (lderrickable) [reply libs/cart/src/effects/cart.effects.ts]: accident, will rebase after approval

File: libs/category/src/effects/category.effects.ts (hunk 8 above)
Comment (lderrickable) [reply libs/category/src/effects/category.effects.ts]: I forgot about this. Still need to do

File: libs/category/src/reducers/category/category.reducer.ts (hunk 7 above)
Comment (lderrickable) [reply libs/category/src/reducers/category/category.reducer.ts]: it can't. typing errors

File: libs/category/src/reducers/category/category.reducer.ts [hunk 11]
```diff
@@ -104,3 +116,96 @@ export function categoryReducer(state = initialState, action: DaffCategoryAction
...
+	toggledFilter: DaffToggledCategoryFilterRequest, 
+	appliedFilters: DaffCategoryFilterRequest[]
+): DaffCategoryFilterRequest[] {
+	return isFilterApplied(toggledFilter, appliedFilters)
+		? removeFilter(toggledFilter, appliedFilters)
+		: addFilter(toggledFilter, appliedFilters)
+}
+
+function isFilterApplied(toggledFilter: DaffToggledCategoryFilterRequest, appliedFilters: DaffCategoryFilterRequest[]): boolean {
+	const filterMatchingName = appliedFilters.find(filter => toggledFilter.name === filter.name);
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/category/src/reducers/category/category.reducer.ts]: removed

File: libs/category/src/reducers/category/category.reducer.ts [hunk 12]
```diff
@@ -104,3 +116,96 @@ export function categoryReducer(state = initialState, action: DaffCategoryAction
...
+	console.log(filterMatchingName)
+	if(!!filterMatchingName && filterMatchingName.type === DaffCategoryFilterType.Match) return filterMatchingName.value === toggledFilter.value;
+	else return !!(<DaffCategoryFilterEqualRequest>filterMatchingName).value.filter(value => value === toggledFilter.value).length;
+}
+
+function removeFilter(toggledFilter: DaffToggledCategoryFilterRequest, appliedFilters: DaffCategoryFilterRequest[])
+	: DaffCategoryFilterRequest[] {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/category/src/reducers/category/category.reducer.ts]: done

File: libs/category/src/reducers/category/category.reducer.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/category/src/reducers/category/category.reducer.ts]: I don't think making subfunctions changes anything about the single responsibility of the parent function, because it still ultimately does those things. But I do think it makes it more readable. Changed

**Comment by lderrickable** [comment]
By the way, I didn't move the tests for the toggle filter functions like i should have. I will in another PR
